### DISCUSSION TITLE
FS.trackingDelegate[callback name] with -sFS_DEBUG will impact performance or not?

### DISCUSSION THREAD

**Discussion by aisnote** [emscripten-core/emscripten#18036] [General]
hi, just curiously **-sFS_DEBUG will impact the performance or not?**

> FS.trackingDelegate[callback name] Users can specify callbacks to receive different filesystem events. This is useful for tracking changes in the filesystem. This requires -sFS_DEBUG.

thanks.

**Comment by tlively** [comment]
Besides the overhead of looking up and running the user callbacks, I wouldn't expect there to be any additional overhead. Note that this feature doesn't have complete coverage of all the file system calls and might go away in the future, so I wouldn't use it for anything more than debugging.

**Reply by sbc100** [reply]
But I imagine `-sFS_DEBUG` could also add other debugging things.. such as tracing.  I would not advise using that setting in production (or any of the XX_DEBUG macros).

I'm not sure what your use case it but perhaps we could find some other way to achieve it, or we could add yet another settings such as a `FS_ENABLE_TRACKING_DELEGATE` (but I very much like to avoid that if possible).

**Reply by sbc100** [reply]
I wonder if you could inject your own interceptors at runtime, using custom JS?  What are you using the tracking callbacks for?

**Comment by aisnote** [comment]
I used to monitor there is new file generated in MEMFS so that I can read that file and post to JS main UI. the scenario is like using FFmpeg.js to generate the TS file continually in MEMFS.

**Reply by sbc100** [reply]
So is is the `close` callback you are specifically interested in?  

I'm afriad I'm not clear on what you mean by "like using FFmpeg.js to generate the TS file continually in MEMFS".  What is a TS file?   Is ffmpeg the program you are running?

**Reply by aisnote** [reply]
ffmpeg.js is WASM module running inside the browser. I call ffmpeg to convert one file like webm to TS files. and ffmpeg will use MEMFS to write the TS files. like:
input: sample.webm.  output: sample0.ts, sample1.ts ...
ffmpeg will use tmp file on writing the content, after one tmp is finished, ffmpeg will call FS.rename from tmp file to ts file. 
So to hook the onCloseFile may an one way to know the TS file is finished.

Hope it clear. Thanks a lot.

**Reply by sbc100** [reply]
I think it would be great if we could find a way to give you the callback you need without requiring `FS.trackingDelegate`.

Can you modify the ffmpeg sources, or do you want a solution that works purely on the JS side?

**Reply by aisnote** [reply]
I can not modify the FFMpeg source code.  Hope EMS can do something on this file operations hook. I know Go lang has this functional to hook FS

**Reply by sbc100** [reply]
I think you should be able to so something like this in your post-js file:

```
// Intercept calls to rename() so we know when the application
// renames a file.
var old_rename = FS.rename;
FS.rename = function(old, new) {
  var rtn = old_rename(old, new);
  callMeOnFileRename(new);
  return rtn;
}
```

**Reply by aisnote** [reply]
let me have a try and very appreciate your help.

**Reply by aisnote** [reply]
thanks a lot @sbc100 . I made it as this.
